File: java/src/leetcode138.java
import java.util.HashMap;

/**
 * @author laegel
 * @version 1.0
 * @since 2026-01-23
 */

class leetcode138 {
    class Node {
        int val;
        Node next;
        Node random;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }
    private HashMap<Node, Node> map = new HashMap<>();

    public Node copyRandomList(Node head) {
        if (head == null) return null;
        if (map.containsKey(head)) return map.get(head);

        Node copy = new Node(head.val);
        map.put(head, copy);
        copy.next = copyRandomList(head.next);
        copy.random = map.get(head.random);
        return copy;
    }

    public Node copyRandomList2(Node head) {
        HashMap<Node, Node> map = new HashMap<>();
        Node cur = head;

        while (cur != null) {
            map.put(cur, new Node(cur.val));
            cur = cur.next;
        }
        cur = head;

        while (cur != null) {
            Node node = map.get(cur);
            node.next = map.get(cur.next);
            node.random = map.get(cur.random);
            cur = cur.next;
        }

        return map.get(head);
    }
}

File: python/leetcode/leetcode138.py

class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def __init__(self):
        self.map = {}

    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None
        if head in self.map:
            return self.map[head]

        copy = Node(head.val)
        self.map[head] = copy
        copy.next = self.copyRandomList(head.next)
        copy.random = self.map.get(head.random)
        return copy

    def copyRandomList2(self, head: 'Optional[Node]') -> 'Optional[Node]':
        map = {None: None}
        cur = head
        while cur:
            copy = Node(cur.val)
            map[cur] = copy
            cur = cur.next
        cur = head

        while cur:
            copy = map.get(cur)
            copy.next = map.get(cur.next)
            copy.random = map.get(cur.random)
            cur = cur.next

        return map[head]
